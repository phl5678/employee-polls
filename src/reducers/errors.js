import { ADD_ERRORS, RESET_ERRORS } from '../actions/errors';

export default function errors(state = [], action) {
  switch (action.type) {
    case ADD_ERRORS:
      return [...state, ...action.errors];
    case RESET_ERRORS:
      return [];
    default:
      return state;
  }
}
