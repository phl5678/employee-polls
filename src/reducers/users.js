import { ANSWER_QUESTION, ADD_QUESTION } from '../actions/questions';
import { RECEIVE_USERS, ADD_AUTH0_USER } from '../actions/users';

export default function users(state = {}, action) {
  switch (action.type) {
    case RECEIVE_USERS:
      return {
        ...state,
        ...action.users,
      };
    case ANSWER_QUESTION:
      return {
        ...state,
        [action.authedUser]: {
          ...state[action.authedUser],
          answers: {
            ...state[action.authedUser].answers,
            [action.qid]: action.answer,
          },
        },
      };
    case ADD_QUESTION:
      return {
        ...state,
        [action.question.author]: {
          ...state[action.question.author],
          questions: state[action.question.author].questions.concat(action.question.id),
        },
      };
    case ADD_AUTH0_USER:
      return {
        ...state,
        [action.user.id]: action.user,
      };
    default:
      return state;
  }
}
