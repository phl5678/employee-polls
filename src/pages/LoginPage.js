import { connect } from 'react-redux';
import { handleUserLogin } from '../actions/authedUser';
import { useEffect, useState } from 'react';
import { useNavigate, useSearchParams } from 'react-router-dom';
import ErrorsList from '../components/ErrorsList';
import { resetErrors } from '../actions/errors';

const LoginPage = ({ dispatch, errors, authedUserID }) => {
  const navigate = useNavigate();
  const [searchParams] = useSearchParams();
  const [id, setID] = useState('');
  const [pw, setPW] = useState('');

  useEffect(() => {
    if (authedUserID !== null) {
      const redirect = searchParams.get('redirect');
      if (redirect !== null) {
        navigate(redirect);
      } else {
        navigate('/home');
      }
    }
  }, [authedUserID, navigate, searchParams]);

  function handleChangeID(e) {
    if (errors && errors.length > 0) dispatch(resetErrors());
    setID(e.target.value);
  }
  function handleChangePW(e) {
    if (errors && errors.length > 0) dispatch(resetErrors());
    setPW(e.target.value);
  }

  function handleSumbit(e) {
    e.preventDefault();
    if (errors && errors.length > 0) dispatch(resetErrors());
    dispatch(handleUserLogin(id, pw));
  }
  return (
    <div className="login center">
      <img src="./hero.png" width="300px" alt="Would You Rather" />
      <h3>Employee Polls</h3>
      <form onSubmit={handleSumbit} data-testid="login-form">
        <div>
          <input
            type="text"
            placeholder="id"
            onChange={handleChangeID}
            value={id}
            data-testid="id-input"
          />
        </div>
        <div>
          <input
            type="password"
            placeholder="password"
            value={pw}
            onChange={handleChangePW}
            data-testid="pw-input"
          />
        </div>
        <ErrorsList errors={errors} />
        <button type="submit" disabled={id === '' || pw === ''} className="btn">
          Log In
        </button>
      </form>
    </div>
  );
};
const mapStateToProps = ({ errors, authedUserID }) => ({ errors, authedUserID });
export default connect(mapStateToProps)(LoginPage);
