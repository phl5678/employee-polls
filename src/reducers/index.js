import { combineReducers } from 'redux';
import authedUser from './authedUser';
import users from './users';
import questions from './questions';
import errors from './errors';
import { loadingBarReducer } from 'react-redux-loading-bar';

export default combineReducers({
  authedUserID: authedUser,
  users,
  questions,
  errors,
  loadingBar: loadingBarReducer,
});
