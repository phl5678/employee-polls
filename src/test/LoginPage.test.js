import { act, render, screen, fireEvent, waitFor } from '@testing-library/react';
import LoginPage from '../pages/LoginPage';
import { BrowserRouter } from 'react-router-dom';
import { createStore } from '@reduxjs/toolkit';
import reducers from '../reducers';
import middleware from '../middleware';
import { Provider } from 'react-redux';

const mockStore = createStore(reducers, middleware);

describe('testing <LoginPage />', () => {
  it('should match the snapshot', () => {
    const comp = render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <LoginPage />
        </BrowserRouter>
      </Provider>
    );
    expect(comp).toMatchSnapshot();
  });

  it('should contain a login form with an ID input field, password input field, and a disabled Submit button.', () => {
    const comp = render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <LoginPage />
        </BrowserRouter>
      </Provider>
    );
    expect(comp.getByTestId('login-form')).toBeInTheDocument();
    expect(comp.getByTestId('id-input')).toBeInTheDocument();
    expect(comp.getByTestId('pw-input')).toBeInTheDocument();
    const button = comp.getByRole('button');
    expect(button).toBeInTheDocument();
    expect(button).toBeDisabled();
  });

  it('should disable submit button when id is empty', () => {
    const user = { id: 'sarahedo', password: 'password123' };
    const comp = render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <LoginPage />
        </BrowserRouter>
      </Provider>
    );

    fireEvent.change(comp.getByTestId('pw-input'), { target: { value: user.password } });
    expect(comp.getByRole('button')).toBeDisabled();
  });

  it('should enable submit button when id and password both not empty.', () => {
    const user = { id: 'sarahedo', password: 'password123' };

    const comp = render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <LoginPage />
        </BrowserRouter>
      </Provider>
    );

    fireEvent.change(comp.getByTestId('id-input'), { target: { value: user.id } });
    fireEvent.change(comp.getByTestId('pw-input'), { target: { value: user.password } });
    expect(comp.getByRole('button')).not.toBeDisabled();
  });

  it('should disaply errors when user not found.', async () => {
    const user = { id: 'test', password: 'test' };
    act(() => {
      render(
        <Provider store={mockStore}>
          <BrowserRouter>
            <LoginPage />
          </BrowserRouter>
        </Provider>
      );
    });

    const form = await waitFor(() => screen.getByTestId('login-form'));
    fireEvent.change(screen.getByTestId('id-input'), { target: { value: user.id } });
    fireEvent.change(screen.getByTestId('pw-input'), { target: { value: user.password } });
    fireEvent.submit(form);

    const error = await waitFor(() => screen.getByTestId('errors-list'));
    expect(error).toBeInTheDocument();
    expect(error.textContent).toEqual('No user found.');
  });

  it('should disaply errors when password not matched.', async () => {
    const user = { id: 'sarahedo', password: 'test' };
    act(() => {
      render(
        <Provider store={mockStore}>
          <BrowserRouter>
            <LoginPage />
          </BrowserRouter>
        </Provider>
      );
    });
    const form = await waitFor(() => screen.getByTestId('login-form'));
    fireEvent.change(screen.getByTestId('id-input'), { target: { value: user.id } });
    fireEvent.change(screen.getByTestId('pw-input'), { target: { value: user.password } });
    fireEvent.submit(form);

    const error = await waitFor(() => screen.getByTestId('errors-list'));
    expect(error).toBeInTheDocument();
    expect(error.textContent).toEqual('Password not matched.');
  });
});
