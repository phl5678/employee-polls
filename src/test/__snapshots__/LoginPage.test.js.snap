// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing <LoginPage /> should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login center"
      >
        <img
          alt="Would You Rather"
          src="./hero.png"
          width="300px"
        />
        <h3>
          Employee Polls
        </h3>
        <form
          data-testid="login-form"
        >
          <div>
            <input
              data-testid="id-input"
              placeholder="id"
              type="text"
              value=""
            />
          </div>
          <div>
            <input
              data-testid="pw-input"
              placeholder="password"
              type="password"
              value=""
            />
          </div>
          <button
            class="btn"
            disabled=""
            type="submit"
          >
            Log In
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login center"
    >
      <img
        alt="Would You Rather"
        src="./hero.png"
        width="300px"
      />
      <h3>
        Employee Polls
      </h3>
      <form
        data-testid="login-form"
      >
        <div>
          <input
            data-testid="id-input"
            placeholder="id"
            type="text"
            value=""
          />
        </div>
        <div>
          <input
            data-testid="pw-input"
            placeholder="password"
            type="password"
            value=""
          />
        </div>
        <button
          class="btn"
          disabled=""
          type="submit"
        >
          Log In
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
